content/en/blog/_posts/2023-09-15-bootstrap-an-air-gapped-cluster-with-kubeadm/index.md

```diff
@@ -0,0 +1,681 @@
+
+In this blog article, I will walk through the process of bootstrapping a Kubernetes
+cluster in an air-gapped lab environment using Fedora Linux and kubeadm.
+
+## The Air Gap VM Setup
+
+A real air-gapped network can take some effort to set up, so for this post, I will use an example VM on a laptop and do some network modifications. Below is the topology:
+
+{{< figure src="airgap-vm.svg" alt="Topology on the host/laptop which shows that connectivity to the internet from the air gap VM is not possible. However, connectivity between the host/laptop and the VM is possible" >}}
+
+### Local Topology
+
+This VM will have its network connectivity disabled but in a way that doesn't shut down the VM's virtual NIC. Instead, its network will be downed by injecting a default route to a dummy interface, making anything internet-hosted unreachable. However, the VM still has a connected route to the bridge interface on the host, which means that network connectivity to the host is still working. This posture means that data can be transferred from the host/laptop to the VM via scp, even with the default route on the VM black-holing all traffic that isn't destined for the local bridge subnet. This type of transfer is analogous to carrying data across the air gap and will be used throughout this post.
+
+Other details about the lab setup:
+
+**VM OS:** Fedora 37
+
+**Kubernetes Version:** v1.27.3
+
+**CNI Plugins Version:** v1.3.0
+
+**CNI Provider and Version:** Flannel v0.22.0
+
+While this single VM lab is a simplified example, the below diagram more approximately shows what a real air-gapped environment could look like:
+
+{{< figure src="example_production_topology.svg" alt="Example production topology which shows 3 control plane Kubernetes nodes and 'n' worker nodes along with a Docker registry in an air-gapped environment.  Additionally shows two workstations, one on each side of the air gap and an IT admin which physically carries the artifacts across." >}}
+
+Note, there is still intentional isolation between the envirnment and the internet.  There are also some things that are not shown in order to keep the diagram simple, for example malware scanning on the secure side of the air gap.
+
+Back to the single VM lab environment.
+
+## Identifying the required software artifacts
+
+I have gone through the trouble of identifying all of the required software components that need to be carried across the air gap in order for this cluster to be stood up:
+
+- Docker (to host an internal container image registry)
+- Containerd
+- libcgroup
+- socat
+- conntrack-tools
+- CNI plugins
+- crictl
+- kubeadm
+- kubelet
+- kubectl and k9s (strictly speaking, these aren't required to bootstrap a cluster but they are handy to interact with one)
+- service systemd file
```

Comment:
BTW, _which_ service?

Reply:
Good catch, it's the `kubelet` systemd file.  Will update outside of the batch edits here.

Reply:
@sftim , specified that this is for the `kubelet.service` systemd file.